File: golang/lib/bulk_command_execution/bulk_command_serializer.go
package bulk_command_execution

import (
	"encoding/json"
	v0_bulk_command_api2 "github.com/kurtosis-tech/kurtosis-client/golang/lib/bulk_command_execution/v0_bulk_command_api"
	"github.com/palantir/stacktrace"
)

const (
	latestSchemaVersion = V0
)

type VersionedBulkCommandsDocument struct {
	SchemaVersion SchemaVersion `json:"schemaVersion"`
}

type serializableBulkCommandsDocument struct {
	VersionedBulkCommandsDocument
	Body	interface{}			`json:"body"`
}

type BulkCommandSerializer struct {}

func NewBulkCommandSerializer() *BulkCommandSerializer {
	return &BulkCommandSerializer{}
}

func (serializer BulkCommandSerializer) Serialize(bulkCommands v0_bulk_command_api2.V0BulkCommands) ([]byte, error) {
	toSerialize := serializableBulkCommandsDocument{
		VersionedBulkCommandsDocument: VersionedBulkCommandsDocument{
			SchemaVersion: latestSchemaVersion,
		},
		Body:                          bulkCommands,
	}
	bytes, err := json.Marshal(toSerialize)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred serializing bulk commands to bytes")
	}
	return bytes, nil
}

File: typescript/lib/bulk_command_execution/bulk_command_serializer.ts
import { V0BulkCommands } from "./v0_bulk_command_api/v0_bulk_commands";
import { SchemaVersion, V0 } from "./bulk_command_schema_version";
import { ok, err, Result } from "neverthrow";

const LATEST_SCHEMA_VERSION = V0;

class VersionedBulkCommandsDocument {
    private readonly schemaVersion: SchemaVersion;
    
    constructor(schemaVersion: SchemaVersion) {
        this.schemaVersion = schemaVersion;
    }
}

class SerializableBulkCommandsDocument extends VersionedBulkCommandsDocument {
    private readonly body: any;
    
    constructor(schemaVersion: SchemaVersion, body: V0BulkCommands) {
        super(schemaVersion);
        this.body = body;
    }
}


class BulkCommandSerializer {

    constructor (){}

    public serialize(bulkCommands: V0BulkCommands): Result<Uint8Array | string, Error> {
        const toSerialize: SerializableBulkCommandsDocument = new SerializableBulkCommandsDocument(LATEST_SCHEMA_VERSION, bulkCommands);
        
        var bytes: string;
        try {
            bytes = JSON.stringify(toSerialize);
        } catch (jsonErr) {
            return err(jsonErr);
        }

        return ok(bytes);
    }
}
